In [ ]:
import sys
!{sys.executable} -m pip install -U python-woc pandas matplotlib

In [5]:
from woc.remote import WocMapsRemote
from tqdm import tqdm
import pandas as pd
# creates the client
woc = WocMapsRemote( base_url="https://worldofcode.org/api/")
# if you got an API key
# woc = WocMapsRemote( base_url="https://worldofcode.org/api/", api_key="woc-XXXXXX-YYYYYY" )

# Now for each of the ten projects you were assigned get commits, e.g.
projects = [
    'Pseudomanifold/Aleph',
    'trichter/rf',
    'evalf/nutils',
    'eXascaleInfolab/bench-vldb20',
    'ashiklom/edr-da',
    'OxfordSKA/OSKAR',
    'ctlab/fgsea',
    'LAMPSPUC/StateSpaceModels.jl',
    'SiEPIC/SiEPIC_EBeam_PDK',
    'PaulHancock/Aegean'
]
list_df_commits = []
for prj in projects:
  # Convert GitHub repository names to WoC V2412 project names
  prj = prj.lower().replace('/','_',2)
  commits = woc.get_values('p2c', prj)
  df = pd.DataFrame(commits, columns=["sha1"])
  df['project'] = prj
  list_df_commits.append(df)
df = pd.concat(list_df_commits)
#perhaps save the list (if no errors) so you do not need to retrieve them again
df.to_csv('df_commits.csv')
df.head(1)

,sha1,project
0,0026ee632a7cd96e05abdbdc7d1fc765aa815aa6,pseudomanifold_aleph


In [8]:
summary = df_commit_data[
    ['project', 'commit', 'author', 'author_time', 'message']
].copy()

summary.columns = [
    'project_wocid',
    'commit_sha1',
    'author',
    'time',
    'commit message'
]

summary.to_csv(
    'dmengeli_project_summary.csv',
    sep=';',
    index=False
)

summary.head()

,project_wocid,commit_sha1,author,time,commit message
0,pseudomanifold_aleph,0026ee632a7cd96e05abdbdc7d1fc765aa815aa6,Bastian Rieck <bastian.rieck@bsse.ethz.ch>,1519028904,Fixed Python integration for Mac OS X\n\nIt is...
1,pseudomanifold_aleph,002d27c8b90928e024aeeafbd9c32d561e0691b2,Bastian Rieck <bastian.rieck@iwr.uni-heidelber...,1497432549,Fixed functor formatting\n
2,pseudomanifold_aleph,00350cd310715806678cfa26a697079ab8cbb793,Bastian Rieck <bastian@rieck.ru>,1485521337,Fixed step function addition\n\nIt is probably...
3,pseudomanifold_aleph,003910ae73dab404d119e889e30e4245a6a610d0,Bastian Rieck <bastian@rieck.ru>,1578045238,Implemented persistence diagram range insertion\n
4,pseudomanifold_aleph,0088bafb190a245f934392d112a2940ec7d694e4,Bastian Rieck <bastian.rieck@iwr.uni-heidelber...,1486737991,Extended GML reader test\n\nLabels are being p...


In [6]:
# If the number of commits its not very large, you can try to get them all at the same time,
# The max batch size is 10
import time
# let us first split df['sha1'] in chunks
chunks = [df['sha1'][x:x+10] for x in range(0, len(df), 10)]
commit_data = []

for chunk in tqdm(chunks): # iterate over the commits
  # res, err = woc.show_content_many('commit',chunk.to_list())

  # commit.tch returns the same data but faster
  res, err = woc.get_values_many('commit.tch',chunk.to_list())
  res = {k: v[0] for k, v in res.items()}  # this conversion is necessary because of the internal implementation

  # to walk around the rate limit
  time.sleep(1)

  if err: # check for errors
    print('Got Errors', err)

  for commit_sha, commit in res.item s():
    # flatten commit objects
    commit_data.append({
          'commit': commit_sha,
          'tree': commit[0],
          'parent': list(commit[1]),
          'author': commit[2][0],
          'author_time': int(commit[2][1]),
          'author_tz': commit[2][2],
          'committer': commit[3][0],
          'committer_time': int(commit[3][1]),
          'committer_tz': commit[3][2],
          'message': commit[4],
    })

df_commit_data = pd.DataFrame(commit_data)
df_commit_data = df_commit_data.merge(df, left_on='commit', right_on='sha1')
df_commit_data.to_csv('df_commit_data.csv', index=False)
df_commit_data.head(2)

 11%|███▉                               | 233/2093 [04:00<32:18,  1.04s/it]

Got Errors {'2b14d054203851cac6e049ecf75208121ccdcffe': 'Key 2b14d054203851cac6e049ecf75208121ccdcffe not found in /da5_fast/All.sha1c/commit_43.tch'}


 11%|███▉                               | 234/2093 [04:01<32:18,  1.04s/it]

Got Errors {'31f9ac970a355af0b0a7aa1c81cee5b5135bd22a': 'Key 31f9ac970a355af0b0a7aa1c81cee5b5135bd22a not found in /da5_fast/All.sha1c/commit_49.tch'}


 11%|███▉                               | 235/2093 [04:02<32:23,  1.05s/it]

Got Errors {'333336d61e77eb6eeae510273306a74f8213d392': 'Key 333336d61e77eb6eeae510273306a74f8213d392 not found in /da5_fast/All.sha1c/commit_51.tch'}


 11%|███▉                               | 238/2093 [04:05<32:06,  1.04s/it]

Got Errors {'433db716fecd2e81379a658c87da160c27f99996': 'Key 433db716fecd2e81379a658c87da160c27f99996 not found in /da5_fast/All.sha1c/commit_67.tch'}


 11%|███▉                               | 239/2093 [04:06<32:06,  1.04s/it]

Got Errors {'4885bdc3faeeb361d38edfd4fb36f13de493b091': 'Key 4885bdc3faeeb361d38edfd4fb36f13de493b091 not found in /da5_fast/All.sha1c/commit_72.tch'}


 12%|████                               | 242/2093 [04:09<32:05,  1.04s/it]

Got Errors {'52b56730563405440d03c38553943da58d4d52bd': 'Key 52b56730563405440d03c38553943da58d4d52bd not found in /da5_fast/All.sha1c/commit_82.tch', '54bde40f9dcfd2d7519a3d7f868184ca9885d686': 'Key 54bde40f9dcfd2d7519a3d7f868184ca9885d686 not found in /da5_fast/All.sha1c/commit_84.tch'}


 12%|████                               | 246/2093 [04:14<32:01,  1.04s/it]

Got Errors {'63ea4f0200dc1e79c510084c2c39b8b86c243d6f': 'Key 63ea4f0200dc1e79c510084c2c39b8b86c243d6f not found in /da5_fast/All.sha1c/commit_99.tch'}


 12%|████▏                              | 249/2093 [04:17<31:55,  1.04s/it]

Got Errors {'7025374cfe9327b714c56f9935aa041244fb866d': 'Key 7025374cfe9327b714c56f9935aa041244fb866d not found in /da5_fast/All.sha1c/commit_112.tch'}


 12%|████▏                              | 250/2093 [04:18<31:57,  1.04s/it]

Got Errors {'7463218a65d6d55e6b087aa87ba1ac3031f42d30': 'Key 7463218a65d6d55e6b087aa87ba1ac3031f42d30 not found in /da5_fast/All.sha1c/commit_116.tch', '76687fa297ce64ce6e734611cb3679a2d787e62d': 'Key 76687fa297ce64ce6e734611cb3679a2d787e62d not found in /da5_fast/All.sha1c/commit_118.tch'}


 12%|████▎                              | 255/2093 [04:23<31:50,  1.04s/it]

Got Errors {'8f2bcfcfea14ad660aa03e748c4ea5bd0e289df7': 'Key 8f2bcfcfea14ad660aa03e748c4ea5bd0e289df7 not found in /da5_fast/All.sha1c/commit_15.tch'}


 12%|████▎                              | 256/2093 [04:24<31:47,  1.04s/it]

Got Errors {'95a500694de7a1904833d3eb205e0f3c5088b715': 'Key 95a500694de7a1904833d3eb205e0f3c5088b715 not found in /da5_fast/All.sha1c/commit_21.tch'}


 12%|████▎                              | 259/2093 [04:27<31:46,  1.04s/it]

Got Errors {'a56e7d8021fdc3a0d20a593049d789c16361070a': 'Key a56e7d8021fdc3a0d20a593049d789c16361070a not found in /da5_fast/All.sha1c/commit_37.tch'}


 13%|████▍                              | 265/2093 [04:33<31:38,  1.04s/it]

Got Errors {'ba17f2603663c7c33c9fad2d76851daf93cc0e9d': 'Key ba17f2603663c7c33c9fad2d76851daf93cc0e9d not found in /da5_fast/All.sha1c/commit_58.tch'}


 13%|████▍                              | 266/2093 [04:34<31:37,  1.04s/it]

Got Errors {'bd69d1ffa22b1883f8e4f0ad99c60fae3c3eb50a': 'Key bd69d1ffa22b1883f8e4f0ad99c60fae3c3eb50a not found in /da5_fast/All.sha1c/commit_61.tch'}


 13%|████▌                              | 270/2093 [04:39<31:35,  1.04s/it]

Got Errors {'d3dd191b669ea2a9520d1ef233d2a7e5b849664e': 'Key d3dd191b669ea2a9520d1ef233d2a7e5b849664e not found in /da5_fast/All.sha1c/commit_83.tch'}


 13%|████▌                              | 271/2093 [04:40<31:33,  1.04s/it]

Got Errors {'d6f9821be3870f30123ac50157eeb529dc689108': 'Key d6f9821be3870f30123ac50157eeb529dc689108 not found in /da5_fast/All.sha1c/commit_86.tch'}


 13%|████▌                              | 275/2093 [04:44<31:28,  1.04s/it]

Got Errors {'ec0df9c78c60c2d0361f5e738053d8e890583946': 'Key ec0df9c78c60c2d0361f5e738053d8e890583946 not found in /da5_fast/All.sha1c/commit_108.tch'}


 53%|██████████████████                | 1113/2093 [19:14<16:54,  1.04s/it]

Got Errors {'3dad1d4758a6ecfba3c750be5d93113f284885a5': 'Key 3dad1d4758a6ecfba3c750be5d93113f284885a5 not found in /da5_fast/All.sha1c/commit_61.tch'}


 65%|█████████████████████▉            | 1352/2093 [23:21<12:53,  1.04s/it]

Got Errors {'9b9d72cfcfc8dc017d0e94aedf380c47ec1a15dd': 'Key 9b9d72cfcfc8dc017d0e94aedf380c47ec1a15dd not found in /da5_fast/All.sha1c/commit_27.tch'}


 67%|██████████████████████▉           | 1411/2093 [24:23<11:54,  1.05s/it]

Got Errors {'5486d8815c89baa3a2d7843a92b020688b54bc82': 'Key 5486d8815c89baa3a2d7843a92b020688b54bc82 not found in /da5_fast/All.sha1c/commit_84.tch'}


 70%|███████████████████████▋          | 1460/2093 [25:14<10:58,  1.04s/it]

Got Errors {'ca416c6821a3c49b6ac99fdc4c7bd50cdf381d4a': 'Key ca416c6821a3c49b6ac99fdc4c7bd50cdf381d4a not found in /da5_fast/All.sha1c/commit_74.tch'}


 88%|██████████████████████████████    | 1852/2093 [32:01<04:10,  1.04s/it]

Got Errors {'00f980c87f9680438f22bf55884ee1cbdb8b8106': 'Key 00f980c87f9680438f22bf55884ee1cbdb8b8106 not found in /da5_fast/All.sha1c/commit_0.tch'}


 89%|██████████████████████████████▏   | 1859/2093 [32:09<04:03,  1.04s/it]

Got Errors {'07d055d1d498f7f19b22535c0a3cfc81fb1d2837': 'Key 07d055d1d498f7f19b22535c0a3cfc81fb1d2837 not found in /da5_fast/All.sha1c/commit_7.tch'}


 90%|██████████████████████████████▌   | 1883/2093 [32:34<03:37,  1.04s/it]

Got Errors {'209a47a24193b20bafc91616198d24f19a91d1d4': 'Key 209a47a24193b20bafc91616198d24f19a91d1d4 not found in /da5_fast/All.sha1c/commit_32.tch'}


 92%|███████████████████████████████▎  | 1924/2093 [33:16<02:55,  1.04s/it]

Got Errors {'4a2ec7605da5a3d0690cf5765c291ccbfd83a1d0': 'Key 4a2ec7605da5a3d0690cf5765c291ccbfd83a1d0 not found in /da5_fast/All.sha1c/commit_74.tch'}


 92%|███████████████████████████████▎  | 1927/2093 [33:19<02:52,  1.04s/it]

Got Errors {'4d7e98699ec9870e50c7dcb6224ea444c330dfef': 'Key 4d7e98699ec9870e50c7dcb6224ea444c330dfef not found in /da5_fast/All.sha1c/commit_77.tch'}


 93%|███████████████████████████████▌  | 1946/2093 [33:39<02:32,  1.04s/it]

Got Errors {'60aecc0a63d27ca4b87ce1088cac78b384eb4398': 'Key 60aecc0a63d27ca4b87ce1088cac78b384eb4398 not found in /da5_fast/All.sha1c/commit_96.tch'}


 93%|███████████████████████████████▋  | 1951/2093 [33:44<02:27,  1.04s/it]

Got Errors {'63e4e2d3edf8d9a6e566bb5a174aded28d81320a': 'Key 63e4e2d3edf8d9a6e566bb5a174aded28d81320a not found in /da5_fast/All.sha1c/commit_99.tch'}


 94%|███████████████████████████████▉  | 1964/2093 [33:58<02:13,  1.04s/it]

Got Errors {'7374998aa2a3d3069600bc17337f93094bef2175': 'Key 7374998aa2a3d3069600bc17337f93094bef2175 not found in /da5_fast/All.sha1c/commit_115.tch'}


 95%|████████████████████████████████▎ | 1988/2093 [34:23<01:48,  1.03s/it]

Got Errors {'8f2fac9cd87841b4f8a9bd276079bc8d7660bdaa': 'Key 8f2fac9cd87841b4f8a9bd276079bc8d7660bdaa not found in /da5_fast/All.sha1c/commit_15.tch'}


 95%|████████████████████████████████▎ | 1989/2093 [34:24<01:47,  1.03s/it]

Got Errors {'8fe5290eac41e0a9893a64f2f14ebba030c7c40d': 'Key 8fe5290eac41e0a9893a64f2f14ebba030c7c40d not found in /da5_fast/All.sha1c/commit_15.tch'}


 96%|████████████████████████████████▍ | 1999/2093 [34:34<01:37,  1.03s/it]

Got Errors {'9b21ba2be8d050910f8fa0a31bfcbd2c232c9889': 'Key 9b21ba2be8d050910f8fa0a31bfcbd2c232c9889 not found in /da5_fast/All.sha1c/commit_27.tch'}


 96%|████████████████████████████████▌ | 2004/2093 [34:39<01:32,  1.04s/it]

Got Errors {'a044ea585ae84c726008298a07380874e2d0fb16': 'Key a044ea585ae84c726008298a07380874e2d0fb16 not found in /da5_fast/All.sha1c/commit_32.tch'}


 96%|████████████████████████████████▊ | 2018/2093 [34:54<01:17,  1.04s/it]

Got Errors {'af5e18aaf212e1beaeb5a5c12624ef129c79933d': 'Key af5e18aaf212e1beaeb5a5c12624ef129c79933d not found in /da5_fast/All.sha1c/commit_47.tch'}


 97%|████████████████████████████████▉ | 2027/2093 [35:03<01:08,  1.03s/it]

Got Errors {'bac984f07cb9ac25d800b771cf725941770ab576': 'Key bac984f07cb9ac25d800b771cf725941770ab576 not found in /da5_fast/All.sha1c/commit_58.tch'}


 97%|████████████████████████████████▉ | 2028/2093 [35:04<01:07,  1.03s/it]

Got Errors {'bc24aa1db42968217dc828c02c9ab9bae466ed0f': 'Key bc24aa1db42968217dc828c02c9ab9bae466ed0f not found in /da5_fast/All.sha1c/commit_60.tch'}


 97%|█████████████████████████████████ | 2039/2093 [35:15<00:55,  1.04s/it]

Got Errors {'c7e93c9949c16efaa901aeb71295bc8b9f173dea': 'Key c7e93c9949c16efaa901aeb71295bc8b9f173dea not found in /da5_fast/All.sha1c/commit_71.tch'}


 99%|█████████████████████████████████▋| 2072/2093 [35:50<00:21,  1.04s/it]

Got Errors {'e96da1bb2e3946d87df39f7f47b46509a786e7a7': 'Key e96da1bb2e3946d87df39f7f47b46509a786e7a7 not found in /da5_fast/All.sha1c/commit_105.tch'}


100%|██████████████████████████████████| 2093/2093 [36:11<00:00,  1.04s/it]


,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,0026ee632a7cd96e05abdbdc7d1fc765aa815aa6,4ca7cb1ff03bcd431873b6e150191cb5e048ebea,[ff4d3decf83f19bec6183365ddc336c44e9bc372],Bastian Rieck <bastian.rieck@bsse.ethz.ch>,1519028904,+0100,Bastian Rieck <bastian.rieck@bsse.ethz.ch>,1519028904,+0100,Fixed Python integration for Mac OS X\n\nIt is...,0026ee632a7cd96e05abdbdc7d1fc765aa815aa6,pseudomanifold_aleph
1,002d27c8b90928e024aeeafbd9c32d561e0691b2,2ffc345f33cc6cc5623feaf9436cd420752c0401,[07d2643da6dfafa71bb6fec111918aca5778dbae],Bastian Rieck <bastian.rieck@iwr.uni-heidelber...,1497432549,+0200,Bastian Rieck <bastian.rieck@iwr.uni-heidelber...,1497432549,+0200,Fixed functor formatting\n,002d27c8b90928e024aeeafbd9c32d561e0691b2,pseudomanifold_aleph


## GitHub Project Statistics

Pseudomanifold/Aleph
Stars: 108
Forks: 15
Last commit date: July 16, 2025

trichter/rf
Stars: 124
Forks: 65
Last commit date: May 5, 2026

evalf/nutils
Stars: 102
Forks: 47
Last commit date: September 21, 2026

eXascaleInfolab/bench-vldb20
Stars: 41
Forks: 15
Last commit date: January 23, 2025

ashiklom/edr-da
Stars: 7
Forks: 4
Last commit date: June 20, 2025

OxfordSKA/OSKAR
Stars: 82
Forks: 46
Last commit date: July 23, 2026

ctlab/fgsea
Stars: 459
Forks: 77
Last commit date: September 22, 2026

LAMPSPUC/StateSpaceModels.jl
Stars: 291
Forks: 30
Last commit date: August 24, 2026

SiEPIC/SiEPIC_EBeam_PDK
Stars: 286
Forks: 184
Last commit date: February 13, 2026

PaulHancock/Aegean
Stars: 52
Forks: 14
Last commit date: July 28, 2025

In [19]:
woc_stats = df_commit_data.groupby('project').agg(
    commits=('commit', 'nunique'),
    authors=('author', 'nunique'),
    max_time=('author_time', 'max'),
    min_time=('author_time', 'min')
)

woc_stats

,commits,authors,max_time,min_time
project,,,,
ashiklom_edr-da,458,5,1750426419,1477337165
ctlab_fgsea,544,33,1762551642,1461616806
evalf_nutils,6587,49,1757000677,1329989907
exascaleinfolab_bench-vldb20,607,10,1737636470,1539276296
lampspuc_statespacemodels.jl,1113,35,1773935433,1516020652
oxfordska_oskar,2743,24,1761294322,1289905814
paulhancock_aegean,2402,23,1753841963,1330396617
pseudomanifold_aleph,2222,10,1752677613,1473339037
siepic_siepic_ebeam_pdk,3658,84,1761227477,1447480391


## World of Code Project Statistics

Project: ashiklom_edr-da
Number of commits: 458
Number of authors: 5
Max time: 1750426419
Min time: 1477337165

Project: ctlab_fgsea
Number of commits: 544
Number of authors: 33
Max time: 1762551642
Min time: 1461616806

Project: evalf_nutils
Number of commits: 6587
Number of authors: 49
Max time: 1757000677
Min time: 1329989907

Project: exascaleinfolab_bench-vldb20
Number of commits: 607
Number of authors: 10
Max time: 1737636470
Min time: 1539276296

Project: lampspuc_statespacemodels.jl
Number of commits: 1113
Number of authors: 35
Max time: 1773935433
Min time: 1516020652

Project: oxfordska_oskar
Number of commits: 2743
Number of authors: 24
Max time: 1761294322
Min time: 1289905814

Project: paulhancock_aegean
Number of commits: 2402
Number of authors: 23
Max time: 1753841963
Min time: 1330396617

Project: pseudomanifold_aleph
Number of commits: 2222
Number of authors: 10
Max time: 1752677613
Min time: 1473339037

Project: siepic_siepic_ebeam_pdk
Number of commits: 3658
Number of authors: 84
Max time: 1761227477
Min time: 1447480391

Project: trichter_rf
Number of commits: 554
Number of authors: 11
Max time: 1771950976
Min time: 1361208896

# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github